### ISSUE TITLE
TGA image loader has buffer overflows

### ISSUE THREAD

**Issue by sfan5** [minetest/irrlicht#236]
Result of fuzzing with [AFL++](https://github.com/AFLplusplus/AFLplusplus). Setup/scripts are in [this branch](https://github.com/minetest/irrlicht/tree/afl).

This resulted in about 20 crashes (probably including duplicates) that look e.g. like this:
<pre>
==329007==ERROR: AddressSanitizer: heap-buffer-overflow on address 0x602000005fd0 at pc 0x5626aacbcc07 bp 0x7ffcd7fe0ea0 sp 0x7ffcd7fe0e98
READ of size 4 at 0x602000005fd0 thread T0
    #0 0x5626aacbcc06 in irr::video::CColorConverter::convert8BitTo16Bit(unsigned char const*, short*, int, int, int const*, int, bool) /var/tmp/irr/source/Irrlicht/CColorConverter.cpp:106:32
    #1 0x5626aacb78e0 in irr::video::CImageLoaderTGA::loadImage(irr::io::IReadFile*) const /var/tmp/irr/source/Irrlicht/CImageLoaderTGA.cpp:189:6
    #2 0x5626aac8e4c6 in irr::video::CNullDriver::createImageFromFile(irr::io::IReadFile*) /var/tmp/irr/source/Irrlicht/CNullDriver.cpp:1155:41
    #3 0x5626aac6f11e in irr::video::CNullDriver::loadTextureFromFile(irr::io::IReadFile*, irr::core::string<char> const&) /var/tmp/irr/source/Irrlicht/CNullDriver.cpp:535:18
    #4 0x5626aac6f11e in irr::video::CNullDriver::getTexture(irr::core::string<char> const&) /var/tmp/irr/source/Irrlicht/CNullDriver.cpp:477:13
    #5 0x5626aa9c2897 in main /var/tmp/irr/examples/LoadTexture/main.cpp:22:23
    #6 0x7f9078227ccf  (/usr/lib/libc.so.6+0x27ccf) (BuildId: 316d0d3666387f0e8fb98773f51aa1801027c5ab)
    #7 0x7f9078227d89 in __libc_start_main (/usr/lib/libc.so.6+0x27d89) (BuildId: 316d0d3666387f0e8fb98773f51aa1801027c5ab)
    #8 0x5626aa883d24 in _start (/var/tmp/irr/bin/Linux/LoadTexture+0xbed24) (BuildId: 9d55f295133956d3)

0x602000005fd0 is located 84 bytes after 12-byte region [0x602000005f70,0x602000005f7c)
freed by thread T0 here:
    #0 0x5626aa9bad5a in operator delete[](void*) (/var/tmp/irr/bin/Linux/LoadTexture+0x1f5d5a) (BuildId: 9d55f295133956d3)
    #1 0x5626aacb6743 in irr::video::CImageLoaderTGA::loadImage(irr::io::IReadFile*) const /var/tmp/irr/source/Irrlicht/CImageLoaderTGA.cpp:139:3
    #2 0x5626aac8e4c6 in irr::video::CNullDriver::createImageFromFile(irr::io::IReadFile*) /var/tmp/irr/source/Irrlicht/CNullDriver.cpp:1155:41

previously allocated by thread T0 here:
    #0 0x5626aa9ba302 in operator new[](unsigned long) (/var/tmp/irr/bin/Linux/LoadTexture+0x1f5302) (BuildId: 9d55f295133956d3)
    #1 0x5626aacb4e01 in irr::video::CImageLoaderTGA::loadImage(irr::io::IReadFile*) const /var/tmp/irr/source/Irrlicht/CImageLoaderTGA.cpp:123:19
    #2 0x5626aac8e4c6 in irr::video::CNullDriver::createImageFromFile(irr::io::IReadFile*) /var/tmp/irr/source/Irrlicht/CNullDriver.cpp:1155:41
</pre>

The following ZIP file contains the ASan output (`.txt`) alongside with the file (`.bin`) that made Irrlicht crash:
[tga_samples.zip](https://github.com/minetest/irrlicht/files/12650712/tga_samples.zip)

**Comment by erlehmann** [comment]
> This resulted in about 20 crashes (probably including duplicates)

Could the duplicates be eliminated using `afl-cmin` and/or `afl-tmin`?

Is this reported to upstream irrlicht or do they have a fix for it maybe?

**Comment by sfan5** [comment]
> Could the duplicates be eliminated using `afl-cmin` and/or `afl-tmin`?

I didn't run either of those on the results, so probably.

> Is this reported to upstream Irrlicht and do they have a fix for it maybe?

It's not and I don't plan on reporting it upstream (currently busy fixing the other bugs AFL found).

**Comment by erlehmann** [comment]
If you are working on AFL stuff anyway, I suggest to run `afl-cmin` (corpus minimization) and `afl-tmin` (test case minimization) on your test cases, in that order then, for everything you find.

Meanwhile I'll dissect the `id:000000.bin` thing.

**Comment by erlehmann** [comment]
Here is what some tools say about `id:000000.bin` renamed to `crash.tga`:

`file`:
```
  Windows Precompiled iNF, version 1.0, InfStyle 1, at 0xff000000, WinDirPath "",, LanguageID 303, at 0x30303, at 0x3
```

`identify`:
```
  crash.tga TGA 8x8 8x8+0+0 8-bit sRGB 4c 112B 0.000u 0:00.000
```

`hachoir-urwid`:
```
  [warn] Skip parser 'TargaFile': (x_min, y_min) is not (0,0)
  Unable to parse file: crash.tga
```

Both GIMP and mtpaint load the file and show the same image. sxiv does not.

Using Imagemagick to convert this to a GIF gives a hint on what could be the issue here:
```
  ; convert crash.tga -auto-orient crash.png
  convert-im6.q16: invalid colormap index `crash.tga' @ error/colormap-private.h/ConstrainColormapIndex/35.
```

**Comment by erlehmann** [comment]
Let's dissect the `crash.tga` byte by byte using some random specification I found on the internet: <http://www.paulbourke.net/dataformats/tga/>.

We start with a hexdump using `hd`:

```
00000000  00 01 01 00 00 04 00 18  00 00 08 14 08 00 08 00  |................|
00000010  08 20 ff 00 00 00 00 ff  01 03 03 03 03 03 03 01  |. ..............|
00000020  20 03 03 03 03 03 03 01  02 02 80 00 00 00 03 01  | ...............|
00000030  02 03 03 03 03 03 03 03  02 03 02 02 02 00 03 00  |................|
00000040  03 03 03 03 03 00 00 00  03 03 03 03 03 00 03 03  |................|
00000050  03 03 03 00 03 00 00 00  00 00 00 00 00 00 54 52  |..............TR|
00000060  55 45 56 49 53 49 4f 4e  2d 58 46 49 4c 45 2e 00  |UEVISION-XFILE..|
00000070
```

The “TRUEVISION-XFILE” footer triggers the TGA parsing in Irrlicht.

Comparing the image against the specification yields:

* Offset 0, 1 byte: 0x00 – no image identification field
* Offset 1, 1 byte: 0x01 – this is a color-mapped (palette) image
* Offset 2, 1 byte: 0x01 – this field always contains 01
* Offset 3, 5 bytes: colormap specification
  * 2 bytes color map origin: 0x00 0x00 – 0
  * 2 bytes colormap length: 0x04 0x00 – 4 palette entries
  * 1 byte colormap bits: 0x18 – we are using 24-bit colormap entries
* Offset 8, 10 bytes: image specification
  * 2 bytes x-origin: 0x00 0x00 – 0
  * 2 bytes y-origin: 0x08 0x14 – 5128 (?), no idea why one would set this
  * 2 bytes width: 0x08 0x00 – 8 pixel wide
  * 2 bytes height: 0x08 0x00 – 8 pixel high
  * 1 byte pixel depth: 0x08 – a pixel is represented by 8 bits (i.e. 1 byte)
  * 1 byte image descriptor: 0x20 = 0b00100000
      * I have no idea why that attribute bit would be set there
      * bit 5 is zero – image origin should be bottomLeft (?)
 * The colormap follows, i.e. 4 × 24 bit colors:
   * color format should be `B8G8R8`
   * 3 bytes: 0xff 0x00 0x00 – blue
   * 3 bytes: 0x00 0x00 0xff – red
   * 3 bytes: 0x01 0x03 0x03 – almost black
   * 3 bytes: 0x03 0x03 0x03 – almost black
* The payload follows, we count colors from zero:
  * 0x03 – almost black
  * 0x01 – red
  * 0x20 – a reference to a 32th color in the colormap!
  * […]
  * 0x80 – a reference to the 128th color in the colormap!

Replacing the two illegal colors yields an image that Imagemagick converts.

**Comment by erlehmann** [comment]
Now the question is, what to do about it. After all, what is crashing here is a conversion from a colormapped image format.

We could very well check each byte in the payload of a colormapped image against the length of the colormap, which would make this case unreachable from this code. But might it be possible to reach the code from the PNG loading code with a colormapped PNG, or maybe any other way?

I am currently trying to figure out the colormap logic and why it is setup that way.

**Comment by erlehmann** [comment]
Besides throwing up on illegal colors, it seems to me that the TGA loader code cheats on color depth for palette images. As far as I understand it, it converts all colors to `ECF_A1R5G5B5`. In the case of the test image, it looks to me like that code reduces the color depth from 8 bit per color to 5 bit per color. If done correctly, the effect is not noticeable to the bare eye (at least not mine) for pixel arts anyway (and I have employed it in the past), but it is a bit cheaty (and probably undesired for palettized image that have an alpha channel).

**Comment by erlehmann** [comment]
Possible approaches for loading the image anyway (like the GIMP and mtPaint):

1. Pad the colormap with black or with the last entry. This is not entirely as easy as it may seem, because that entry must match the color depth of the picture.
2. Change the color conversion code so that an illegal entry referrs to black or the last colormap entry

Possible approaches for preventing loading a malformed image in the TGA encoder (I would prefer that):

3. Validate the input against a TGA grammar. Lots of work and input is context-sensitive, I doubt anyone will do this anytime soon. It is not done for other file formats either.
4. Check that each byte of a payload iof a color-mapped image with pixel depth 8 is smaller (or equal? there is a chance for an off-by-one error here and I am sleepy) than `header.ColorMapLength` and exit if it is, before attempting conversion.
5. Catch the error in the color conversion code.

I think that option number 4 is the easiest to implement. @sfan5 what do you think?

**Comment by erlehmann** [comment]
I think at least the BMP loader might have the same problem, as it also does color conversions.

Judging from the assumptions made in the code, I suggest all code that converts from and to color formats needs a lot of test cases.

**Comment by erlehmann** [comment]
Reported to Irrlicht: https://irrlicht.sourceforge.io/forum/viewtopic.php?t=52925

My idea behind this is lazyness: If Irrlicht gets a fix, it is less work for Minetest.

**Comment by sfan5** [comment]
> I think that option number 4 is the easiest to implement.

1 or 4

**Comment by numberZero** [comment]
> Pad the colormap with black or with the last entry. This is not entirely as easy as it may seem, **because that entry must match the color depth of the picture.**

Why should it? Just pad with enough zeroes for all color depths.

**Comment by erlehmann** [comment]
>>  Pad the colormap with black or with the last entry. This is not entirely as easy as it may seem, **because that entry must match the color depth of the picture.**
>
> Why should it? Just pad with enough zeroes for all color depths.

How many zeroes exactly would you use? I have not thought much about it, but given that the size of the colormap is a 16 bit integer and one can have 4 bytes per color, would you suggest that the decoder should allocate and zero (2¹⁶ - 1) × 4 = 262140 bytes for each colormapped image that needs to be loaded?

Btw, I fixed `tga_encoder` mod so that buffer-overflowing invalid TGA images can no longer be encoded:
https://git.minetest.land/erlehmann/tga_encoder/commit/bb68ceb38d5509e69f7d348f74afc0eca77f062c

**Comment by numberZero** [comment]
> How many zeroes exactly would you use? I have not thought much about it, but given that the size of the colormap is a 16 bit integer and one can have 4 bytes per color, would you suggest that the decoder should allocate and zero (2¹⁶ - 1) × 4 = 262140 bytes for each colormapped image that needs to be loaded?

That’s 1/4 MiB, or 64 pages. Tolerable as it’s freed after decoding, if allocation is fast.

But wait, the palette is *only* used for 8 bpp images. So by no way could indices above 255 be accessed during conversion. Thus 1 KiB. And palettes larger than 256 entries can be rejected outright as invalid.

**Comment by erlehmann** [comment]
> That’s 1/4 MiB, or 64 pages. Tolerable as it’s freed after decoding, if allocation is fast.

How could I test how fast allocating and zero-ing memory is on my hardware?

I still think that allocating memory instead of checking for validity is a bad idea on principle.

I think this because it allows decoding invalid images. One may have to *keep* decoding them as a consequence of it.

> But wait, the palette is only used for 8 bpp images. So by no way could indices above 255 be accessed during conversion. Thus 1 KiB.

Thanks. As I said, I did not think too much about it!

> And palettes larger than 256 entries can be rejected outright as invalid.

They are not really invalid, but I get your point here. When you need 2 bytes per pixel, then A1R5G5B5 is probably a better format, unless you *really* need that alpha channel bit-depth – and Minetest definitely does not provide it for paletted images anyway.

**Comment by grorp** [comment]
Looks like the Irrlicht maintainer fixed some or all of this: https://irrlicht.sourceforge.io/forum/viewtopic.php?p=307192#p307192

**Comment by sfan5** [comment]
The recent commits are in [this branch](https://github.com/minetest/irrlicht/commits/svn-trunk), but we haven't picked commits from there for a while so might not work without conflicts.

**Comment by sfan5** [comment]
After the latest commits there is at least one issue remaining, e.g.:
```
=================================================================
==48098==ERROR: AddressSanitizer: heap-buffer-overflow on address 0x611000000640 at pc 0x559359d14df5 bp 0x7ffdf2074230 sp 0x7ffdf20739f0
WRITE of size 12 at 0x611000000640 thread T0
    #0 0x559359d14df4 in __interceptor_fread.part.0 (/var/tmp/irr/bin/Linux/LoadTexture+0x111df4) (BuildId: 7e2e92a708c7b2e0)
    #1 0x55935a0f9e1e in irr::video::CImageLoaderTGA::loadCompressedImage(irr::io::IReadFile*, irr::video::STGAHeader const&) const /var/tmp/irr/source/Irrlicht/CImageLoaderTGA.cpp:48:10
    #2 0x55935a0f60c0 in irr::video::CImageLoaderTGA::loadImage(irr::io::IReadFile*) const /var/tmp/irr/source/Irrlicht/CImageLoaderTGA.cpp:173:10
    #3 0x55935a0cd8d6 in irr::video::CNullDriver::createImageFromFile(irr::io::IReadFile*) /var/tmp/irr/source/Irrlicht/CNullDriver.cpp:1155:41
    #4 0x55935a0ae56e in irr::video::CNullDriver::loadTextureFromFile(irr::io::IReadFile*, irr::core::string<char> const&) /var/tmp/irr/source/Irrlicht/CNullDriver.cpp:535:18
    #5 0x55935a0ae56e in irr::video::CNullDriver::getTexture(irr::core::string<char> const&) /var/tmp/irr/source/Irrlicht/CNullDriver.cpp:477:13
    #6 0x559359e014a7 in main /var/tmp/irr/examples/LoadTexture/main.cpp:22:23
    #7 0x7f8b6cb58ccf  (/usr/lib/libc.so.6+0x27ccf) (BuildId: 8bfe03f6bf9b6a6e2591babd0bbc266837d8f658)
    #8 0x7f8b6cb58d89 in __libc_start_main (/usr/lib/libc.so.6+0x27d89) (BuildId: 8bfe03f6bf9b6a6e2591babd0bbc266837d8f658)
    #9 0x559359cc2934 in _start (/var/tmp/irr/bin/Linux/LoadTexture+0xbf934) (BuildId: 7e2e92a708c7b2e0)

0x611000000640 is located 0 bytes after 256-byte region [0x611000000540,0x611000000640)
allocated by thread T0 here:
    #0 0x559359df8f12 in operator new[](unsigned long) (/var/tmp/irr/bin/Linux/LoadTexture+0x1f5f12) (BuildId: 7e2e92a708c7b2e0)
    #1 0x55935a0f9c16 in irr::video::CImageLoaderTGA::loadCompressedImage(irr::io::IReadFile*, irr::video::STGAHeader const&) const /var/tmp/irr/source/Irrlicht/CImageLoaderTGA.cpp:36:13
    #2 0x55935a0f60c0 in irr::video::CImageLoaderTGA::loadImage(irr::io::IReadFile*) const /var/tmp/irr/source/Irrlicht/CImageLoaderTGA.cpp:173:10
    #3 0x55935a0cd8d6 in irr::video::CNullDriver::createImageFromFile(irr::io::IReadFile*) /var/tmp/irr/source/Irrlicht/CNullDriver.cpp:1155:41
```

new results: 
[tga_samples2.zip](https://github.com/minetest/irrlicht/files/12921024/tga_samples2.zip)

**Comment by mzeilfelder** [comment]
Thanks, yeah I missed that I had to add checks for the file read as well.
Fixed in 1.8 branch: https://sourceforge.net/p/irrlicht/code/6560 (also switched a few variables from s32 to u32, should probably have been 2 patches, but was too lazy)
Same patch merged to trunk in https://sourceforge.net/p/irrlicht/code/6561

**Comment by sfan5** [comment]
Looks good, I found no new issues during a quick fuzzer run.
